docs: revise and expand `ImmutableDict` docstring.

rfritz opened the pull request:
I had difficulty using the ImmutableDict type based on the documentation. After studying the code and reviewing advice on the Discourse site, I wrote this. This is the second version; I have made it match the style of the existing dictionary documentation and removed the tutorial content.

rfritz committed:
collections.md: add ImmutableDict to list of dictionary types

rfritz committed:
dict.jl: revise and expand ImmutableDict docstring for clarity

giordano commented:
Thanks for the contribution, but can we please have a more descriptive title? The current one is rather cryptic.

rfritz commented:
Ooops! That was pretty obscure. I've updated the commit message on dict.jl following the procedure (and wincing at the warnings) given in "Changing a commit message." Will that serve?

giordano commented:
Ok, I changed the title of the PR based on commit message.  As reference for next time, you can do it by clicking on the "Edit" button in the top right corner of the page.

rfritz commented:
oh, duh. thanks for the correction and information.

MasonProtter commented on base/dict.jl:
```diff
@@ -776,20 +776,60 @@ struct ImmutableDict{K,V} <: AbstractDict{K,V}
 
 """
-    ImmutableDict
+    ImmutableDict(key=>value, key=>value, ...)
```
It's maybe confusing to reuse the same key and value names, I'd instead write it as 

```suggestion
    ImmutableDict(key1=>value1, key2=>value2, ...)
```
or something similar

MasonProtter commented on base/dict.jl:
```diff
@@ -776,20 +776,60 @@ struct ImmutableDict{K,V} <: AbstractDict{K,V}
+linked list.
 
-Create a new entry in the `ImmutableDict` for a `key => value` pair
+    Base.ImmutableDict(key=>value, key=>value, ...)
```
```suggestion
```
This line seems to be a duplicate?

MasonProtter commented on base/dict.jl:
```diff
@@ -776,20 +776,60 @@ struct ImmutableDict{K,V} <: AbstractDict{K,V}
-Create a new entry in the `ImmutableDict` for a `key => value` pair
+    Base.ImmutableDict(key=>value, key=>value, ...)
+
+`Base.ImmutableDict(key=>value, key=>value, ...)` constructs a linked
```
```suggestion
`Base.ImmutableDict(key1=>value1, key2=>value2, ...)` constructs a linked
```

MasonProtter commented on base/dict.jl:
```diff
@@ -776,20 +776,60 @@ struct ImmutableDict{K,V} <: AbstractDict{K,V}
+`Base.ImmutableDict{K,V}()` constructs an empty list with keys of type
+K and values of type V.
+
+    Base.ImmutableDict(imdict, key=>value, key=>value, ...)
```
```suggestion
    Base.ImmutableDict(imdict, key1=>value1, key2=>value2, ...)
```

MasonProtter commented on base/dict.jl:
```diff
@@ -776,20 +776,60 @@ struct ImmutableDict{K,V} <: AbstractDict{K,V}
+
+    Base.ImmutableDict(imdict, key=>value, key=>value, ...)
+
+`Base.ImmutableDict(imdict, key=>value, key=>value, ...)` constructs a
```
```suggestion
`Base.ImmutableDict(imdict, key1=>value1, key2=>value2, ...)` constructs a
```

MasonProtter reviewed (commented):
Just some minor comments / tweaks

rfritz committed:
Switch key=>value, key=>value to key1=>value1, key2=>value2 for clarity

Co-authored-by: Mason Protter <mason.protter@icloud.com>

LilithHafner commented:
Thanks for continuing to work on this, @rfritz! FYI, in the future you can continue using the [previous pr](https://github.com/JuliaLang/julia/pull/58360) and don't have to open a new one, that way all the conversation stays in the same place. We can keep working here now, though :)

MasonProtter commented on base/dict.jl:
```diff
@@ -776,20 +776,59 @@ struct ImmutableDict{K,V} <: AbstractDict{K,V}
 
 """
-    ImmutableDict
+    ImmutableDict(key=>value, key=>value, ...)
```
```suggestion
    ImmutableDict(key1=>value1, key2=>value2, ...)
```
I think this one got missed

rfritz committed:
Switch a last key=>value to numbered keys and values.

Co-authored-by: Mason Protter <mason.protter@icloud.com>

rfritz commented:
Is there anything else I need to do on this? Or is it just a matter of waiting until it's picked up in a release?

rfritz committed:
collections.md: remove a superfluous space

Co-authored-by: Mosè Giordano <765740+giordano@users.noreply.github.com>

rfritz committed:
Merge branch 'JuliaLang:master' into imdictdoc

fingolfin reviewed (commented):
Just discovered this PR. Thanks for working on this, but I do think there is some more work needed. Sorry for taking so long to get you feedback :-(

rfritz committed:
Reorganize ImmutableDict docs for constructor signatures

Move the constructor signature descriptions to be just above
the documentation for the constructor.

rfritz committed:
dict.jl: convert ImmutableDict examples to doctest

rfritz commented on base/dict.jl:
```diff
@@ -778,20 +778,59 @@ struct ImmutableDict{K,V} <: AbstractDict{K,V}
+               viewparameters, "type"=>"perspective", "direction"=>(0, 0, 1))
+    Base.ImmutableDict{String, Any} with 2 entries:
+      "direction" => (0, 0, 1)
+      "type"      => "perspective"
```
I've wrapped the examples in a `jldoctest` block; I hope I've got them right.
